zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -380,1147 +220,479 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
-                        0,
-                        F::zero(),
+                let mut rpi_rlc_acc_cell = None;
+                for (annotation, field_type, field_bytes) in public_data.assignments() {
+                    let cells = self.assign_pi_field(
+                        region,
+                        &mut offset,
+                        annotation,
+                        &field_bytes,
+                        &mut rpi_rlc_acc,
+                        &mut rpi_len_acc,
+                        challenges,
                         false,
-                        F::zero(),
-                        &mut rpi_vals,
                     )?;
-                    offset += 1;
+                    if field_type == FieldType::BlockHash {
+                        region.constrain_equal(
+                            block_table_hash_cell.cell(),
+                            cells[RPI_CELL_IDX].cell(),
+                        )?;
+                    }
+                    rpi_rlc_acc_cell = Some(cells[RPI_RLC_ACC_CELL_IDX].clone());
                 }
-                // NOTE: we add this empty row so as to pass mock prover's check
-                //      otherwise it will emit CellNotAssigned Error
-                let tx_table_len = TX_LEN * self.max_txs + 1;
-                self.assign_tx_empty_row(&mut region, tx_table_len + offset)?;
-
-                let (origin_txs_hash_hi_cell, origin_txs_hash_lo_cell, txs_rlc_acc) =
-                    self.assign_txs(&mut region, public_data, challenges, rpi_vals)?;
-                // assert two txs hash are equal
-                region.constrain_equal(txs_hash_hi_cell.cell(), origin_txs_hash_hi_cell.cell())?;
-                region.constrain_equal(txs_hash_lo_cell.cell(), origin_txs_hash_lo_cell.cell())?;
-                Ok((
-                    [block_hash_hi_cell, block_hash_lo_cell],
-                    txs_rlc_acc,
-                    block_rlc_acc,
-                ))
+
+                // input_rlc in self.rpi
+                // input_len in self.rpi_len_acc
+                // output_rlc in self.rpi_rlc_acc
+                let keccak_row = offset;
+                let rpi_rlc_acc_cell = rpi_rlc_acc_cell.unwrap();
+                rpi_rlc_acc_cell.copy_advice(
+                    || "keccak(rpi)_input",
+                    region,
+                    self.rpi,
+                    keccak_row,
+                )?;
+                let keccak = public_data.get_pi();
+                let keccak_rlc =
+                    keccak
+                        .to_fixed_bytes()
+                        .iter()
+                        .fold(Value::known(F::zero()), |acc, byte| {
+                            acc.zip(challenges.keccak_input())
+                                .and_then(|(acc, randomness)| {
+                                    Value::known(acc * randomness + F::from(*byte as u64))
+                                })
+                        });
```

Comment:
`rlc::value()`?

Reply:
I think `rlc::value()` is bit little weird, it needs reverse the input bytes, so I need reverse `keccak` before passing into it.

Reply:
Still, reversing the input bytes and calling rlc::value() is much more readable and clear. Re-implementing the rlc encoding logic, while quite simple, still a bit tricky!

Reply:
I have used the `rlc::value` for reducing code